fvm/handler/programs.go

```diff
@@ -16,80 +22,170 @@ type TransactionPrograms interface {
 	return nil
 }
 
-func (h *ProgramsHandler) Get(location common.Location) (*interpreter.Program, bool) {
+func (programs *Programs) get(
+	location common.Location,
+) (
+	*interpreter.Program,
+	bool,
+) {
 	// ignore empty locations
 	if location == nil {
 		return nil, false
 	}
 
+	program, state, has := programs.transactionPrograms.Get(location)
+	if has {
+		if state != nil {
+			err := programs.stateTransaction.AttachAndCommitParseRestricted(
+				state)
+			if err != nil {
+				panic(fmt.Sprintf(
+					"merge error while getting program, panic: %s",
+					err))
+			}
+		}
+		return program, true
+	}
+
 	address, ok := location.(common.AddressLocation)
+	if ok {
+		// Address location program is reusable across transactions.  Create
+		// a nested transaction here in order to capture the states read to
+		// parse the program.
+		_, err := programs.stateTransaction.BeginParseRestrictedNestedTransaction(
+			address)
+		if err != nil {
+			panic(err)
+		}
+	}
+	return nil, false
+}
 
-	// program cache track only for AddressLocation
-	if !ok {
-		prog, ok := h.nonAddressPrograms[location.ID()]
-		return prog, ok
+func (programs *Programs) GetProgram(
+	location common.Location,
+) (
+	*interpreter.Program,
+	error,
+) {
+	defer programs.tracer.StartSpanFromRoot(trace.FVMEnvGetProgram).End()
+
+	err := programs.meter.MeterComputation(environment.ComputationKindGetProgram, 1)
+	if err != nil {
+		return nil, fmt.Errorf("get program failed: %w", err)
 	}
 
-	program, state, has := h.Programs.Get(address)
-	if has {
-		err := h.masterState.AttachAndCommitParseRestricted(state)
-		if err != nil {
-			panic(fmt.Sprintf("merge error while getting program, panic: %s", err))
+	if addressLocation, ok := location.(common.AddressLocation); ok {
+		address := flow.Address(addressLocation.Address)
+
+		freezeError := programs.accounts.CheckAccountNotFrozen(address)
+		if freezeError != nil {
+			return nil, fmt.Errorf("get program failed: %w", freezeError)
 		}
-		return program, true
 	}
 
-	_, err := h.masterState.BeginParseRestrictedNestedTransaction(address)
+	program, has := programs.get(location)
+	if has {
+		return program, nil
+	}
+
+	return nil, nil
+}
+
+func (programs *Programs) SetProgram(
+	location common.Location,
+	program *interpreter.Program,
+) error {
+	defer programs.tracer.StartSpanFromRoot(trace.FVMEnvSetProgram).End()
+
+	err := programs.meter.MeterComputation(environment.ComputationKindSetProgram, 1)
 	if err != nil {
-		panic(err)
+		return fmt.Errorf("set program failed: %w", err)
 	}
-	return nil, false
+
+	err = programs.set(location, program)
+	if err != nil {
+		return fmt.Errorf("set program failed: %w", err)
+	}
+	return nil
+}
+
+func (programs *Programs) DecodeArgument(
+	bytes []byte,
+	_ cadence.Type,
+) (
+	cadence.Value,
+	error,
+) {
+	defer programs.tracer.StartExtensiveTracingSpanFromRoot(
+		trace.FVMEnvDecodeArgument).End()
+
+	v, err := jsoncdc.Decode(programs.meter, bytes)
```

Comment:
note that this originally used the full environment instead of just the meter object as the first argument.